### ISSUE TITLE
Symfony authenticator's no longer called with anonymous: lazy

### ISSUE THREAD

**Issue by weaverryan** [symfony/symfony#34614]
**Symfony version(s) affected**: 4.4.0

**Description**  
Until now, if you have a Guard authenticator, its `supports()` method will be called on every request. However, with `anonymous: lazy` under your firewall, that no longer happens.

Maybe that's intended? If it is, I think `anonymous: lazy` can't be used as the default config in the `security.yaml` recipe, as it creates a very different experience.

**How to reproduce**  
Reproducer: https://github.com/weaverryan/symfony-anonymous-lazy-reproducer

Basically, create an authenticator on a new project and access any URL. It will not be executed (because `anonymous: lazy` is the default setting in the recipe).

**Possible Solution**  
I'm not sure if this is a bug with `anonymous: lazy` or if we should revert the recipe back to `anonymous: true`.

**Additional context**  
Caught by MakerBundle tests :)

**Comment by nicolas-grekas** [comment]
What's the issue in terms of real-world use case? Here is the answer about lazy: disabling laziness will make HTTP caches incompatible with firewalls again. Not sure this is what we want.

**Comment by weaverryan** [comment]
> What's the issue in terms of real-world use case? 

I think a normal login form won't work anymore. When the user does a `POST /login_check`, if I understand correctly, there is nothing that will cause the authenticator to be called. The request will continue to the controller, which will just render the login form again.

**Comment by nicolas-grekas** [comment]
> The request will continue to the controller, which will just render the login form again.

Interesting. Does the reproducer above showcase this? Could you update it to do so otherwise please?

**Comment by weaverryan** [comment]
It does:

```
git@github.com:weaverryan/symfony-anonymous-lazy-reproducer.git
cd symfony-anonymous-lazy-reproducer
composer install
symfony serve -d
curl https://127.0.0.1:8000
```

When you try this, you'll hit `DefaultController::index()` (the homepage controller). But you should NOT, because `AppCustomAuthenticator::supports()` has a `die()` statement in it. Change to `anonymous: true` and try the `curl` request again and you'll hit the `supports()` method in the authenticator.

Let me know if it's not clear - I'm not very educated about the `anonymous: lazy` yet - so apologies.

**Comment by nicolas-grekas** [comment]
I'd prefer a real guard: the die is just a symptom, I'd prefer not fixing the symptom without understanding the full lifecycle.

**Comment by dunglas** [comment]
I've the same problem in an app I'm upgrading:

```yaml
            json_login:
                check_path: login
                success_handler: Dunglas\UserBundle\Handler\JsonLoginSuccessHandler
```

```php
    public function testBadLogin(): void
    {
        $client = static::createClient();
        $response = $client->request('POST',
            '/login',
            (new HttpOptions())->setJson([
                'username' => 'dunglas@gmail.com',
                'password' => 'badPassword',
            ])->toArray()
        );
        $this->assertSame(401, $response->getStatusCode()); // Now, it's a 404
    }
```

**Comment by weaverryan** [comment]
Fair enough :). Let's try this again:

```
git clone git@github.com:weaverryan/symfony-anonymous-lazy-reproducer.git
cd symfony-anonymous-lazy-reproducer
composer install
symfony serve -d
```

This now has a standard login form (from `make:auth`). Go to `/login` and fill out the form. You will see the login form again with no errors. This is because the `LoginFormAuthenticator` is never called.

If you (as you know) even do *one* thing on your page to access auth (even an `app.user` reference in the template), it WILL trigger the system. So now I understand that this is the desired behavior of the `anonymous: lazy` system. I'm still worried the behavior is too strange. Here's another problematic example: OAuth. When an OAuth server redirects back to my site, I typically have an authenticator for that URL that fetches the access token, does work, then redirects. That authenticator will no longer be called.
